Replay parsing performance improvements

Rogerstalker opened the pull request:
Fixes #250
Affects https://github.com/FAForever/downlords-faf-client/issues/3270 and PR https://github.com/FAForever/downlords-faf-client/pull/3394
### Description

Switched parser logic to ByteBuffer from LittleEndianDataInputStream for more direct and performant access to the originally read replay bytes and easier reusability during parsing

Eliminated double reading of stream and multiple smaller tokenContent stream creation for tokenization and parsing by not preparing separate tokenContent byte chucks and reusing the originally created ByteBuffer

Replaced tokenContent and tokenSize with token limit, which stores the read position limit in the ByteBuffer for the corresponding token

Added buffer limiting to parsing to preserve robustness coming from tokenization

Improved performance of readString methods by replacing the ByteArrayOutputStream and double accessing the ByteBuffer for string ending search and extracting string bytes

Implemented getUnsignedByte util for ByteBuffer

Removed unused `commandsPerMinuteByPlayer` event interpret calculation.

### Measurements
I measured the performance of the parser with the help of VisualVM and the ReplayLoaderTableParserTest test class. CPU: AMD Ryzen 9 5950x. I parsed all my local replays (822 at the time of measurements) with the following code:
```
    sleep(10000);

    File folder = new File("c:/Users/Devastator/Downloads/replays/");
    File[] listOfFiles = folder.listFiles();

    var count = new AtomicInteger();
    for (final var filePath : listOfFiles) {
      System.out.println(count.incrementAndGet());
      new ReplayDataParser(filePath.toPath(), objectMapper);
    }
```

Except for the last parallel test:
```
    sleep(10000);

    File folder = new File("c:/Users/Devastator/Downloads/replays/");
    File[] listOfFiles = folder.listFiles();

    var count = new AtomicInteger();

    ExecutorService EXEC = Executors.newFixedThreadPool(4);
    List<Callable<ReplayDataParser>> tasks = new ArrayList<Callable<ReplayDataParser>>();

    for (final var filePath : listOfFiles) {
      Callable<ReplayDataParser> c = new Callable<ReplayDataParser>() {
        @Override
        public ReplayDataParser call() throws Exception {
          System.out.println(count.incrementAndGet());
          return new ReplayDataParser(filePath.toPath(), objectMapper);
        }
      };
      tasks.add(c);
    }
    List<Future<ReplayDataParser>> results = EXEC.invokeAll(tasks);
```
The sleep call is necessary for having time to attach the profiler to the process and start sampling.

See attached measurements:
[Common Replay Parser Profiling.zip](https://github.com/user-attachments/files/22199360/Common.Replay.Parser.Profiling.zip)

Baseline
**Original.nps:** com.faforever.commons.replay.ReplayDataParser.<init> ()	32 699 ms

Replaced LittleEndianDataInputStream with ByteBuffer just in the tokenizer and used `System.arraycopy()` to extract `TokenContent` bytes
**ByteBuffer tokenizer.nps:** com.faforever.commons.replay.ReplayDataParser.<init> ()	27 798 ms

Used `ByteBuffer `instead of the `LittleEndianDataInputStream ` everywhere throughout the parsing process.
**ByteBuffer full.nps:** com.faforever.commons.replay.ReplayDataParser.<init> ()	22 499 ms

Reused the initially created `ByteBuffer` everywhere to avoid buffer creation for individual tokens.
**ByteBuffer full final.nps:** com.faforever.commons.replay.ReplayDataParser.<init> ()	17 898 ms

Used this measurement to determine how parallelization of different replay parsing is scaling. By not limiting the number of  used threads I ran into out of memory exceptions due to heap constraints, so I elected to limit them according to my measurements and saw the best performance with a limit of 4. Up to 4 threads there were measurable performance improvements, however more than 4 threads made the parsing actually slower.
**ByteBuffer full final parallel.nps:** com.faforever.commons.replay.ReplayLoaderTableParserTest.ExecuteParallel ()	13 299 ms

Rogerstalker committed:
Switched parser logic to ByteBuffer from LittleEndianDataInputStream for more direct and performant access to the originally read replay bytes and easier reusability during parsing
Eliminated double reading of stream and multiple smaller tokenContent stream creation for tokenization and parsing by not preparing separate tokenContent byte chucks and reusing the originally created ByteBuffer
Replaced tokenContent and tokenSize with token limit, which stores the read position limit in the ByteBuffer for the corresponding token
Added buffer limiting to parsing to preserve robustness coming from tokenization
Improved performance of readString methods by replacing the ByteArrayOutputStream and double accessing the ByteBuffer for string ending search and extracting string bytes
Implemented getUnsignedByte util for ByteBuffer

Sheikah45 commented on data/src/main/java/com/faforever/commons/replay/ReplayDataParser.java:
```diff
@@ -364,10 +359,13 @@ private void interpretEvents(List<Event> events) {
+    LuaData msg;
+    LuaData from;
+    LuaData sender;
+    if ((msg = lua.value().get("Msg")) != null && (from = lua.value().get("From")) != null && (sender = lua.value().get("Sender")) != null) {
```
These null checks are unnecessary since instanceOf is a null check as well

Rogerstalker replied:
removed

Sheikah45 commented:
I am not sure that I like the replacement of the tokenContent with a pointer inside of the original buffer. Part of the purpose of introducing the token was so that it could be standalone and easier to understand in it's own context and handle. I don't think this level of optimization is needed at the cost of the readability and parseability. 

I can see the utility of just using the ByteBuffers

Rogerstalker committed:
Removed unnecessary null checks from parseGiveResourcesToPlayer

Rogerstalker commented:
> I am not sure that I like the replacement of the tokenContent with a pointer inside of the original buffer. Part of the purpose of introducing the token was so that it could be standalone and easier to understand in it's own context and handle. I don't think this level of optimization is needed at the cost of the readability and parseability.
> 
> I can see the utility of just using the ByteBuffers

I am unfamiliar with the edge cases, that could occur during parsing, so there could be some merit to extracting the token content during tokenization, however this change alone accounted for around a 25% performance improvement in a feature that is frequently used and its performance was noticably lacking, so I thought it is worth it.

Brutus5000 commented:
This is a very interesting contribution and I was not aware of ByteBuffers or even its capability to go LittleEndian. I always wondered why we need Guava to do such a basic thing, well we were using the wrong thing as it seems.

I think the benefits are worth breaking the compatibility here. But we also need to adapt the moderator client later on.

So if we are breaking compatibility anyway we should go all in 😛  Since you are investigating and measuring I have a request: can you check if this also makes sense to apply to loading the data? There we still copy around data and that could be avoided.

This is a suggestion of ChatGPT, I didn't check it

```java
private void readReplayData(Path replayFile) throws IOException, CompressorException {
    // Read everything into a single byte[]
    byte[] allReplayData = Files.readAllBytes(replayFile);
    ByteBuffer buffer = ByteBuffer.wrap(allReplayData);

    // Find header end (still using your helper, which works on byte[])
    int headerEnd = findReplayHeaderEnd(allReplayData);

    // Parse metadata directly from a slice (no Arrays.copyOf)
    ByteBuffer headerSlice = buffer.duplicate();
    headerSlice.limit(headerEnd); // [0, headerEnd)
    metadata = objectMapper.readValue(
        new String(headerSlice.array(), headerSlice.position(), headerSlice.remaining(), StandardCharsets.UTF_8),
        ReplayMetadata.class
    );

    // Slice for the compressed data (no Arrays.copyOfRange)
    ByteBuffer dataSlice = buffer.duplicate();
    dataSlice.position(headerEnd + 1);
    dataSlice = dataSlice.slice(); // starts at headerEnd+1

    // If decompress() can be updated to accept ByteBuffer, pass it directly
    // Otherwise, fall back to array without copying
    if (decompressAcceptsByteBuffer()) {
        data = decompress(dataSlice, metadata);
    } else {
        data = decompress(dataSlice.array(), dataSlice.arrayOffset() + dataSlice.position(), dataSlice.remaining(), metadata);
    }
}
```

Rogerstalker committed:
Improved qUncompress performance
- Eliminated unnecessary array copying by reusing the inputBuffer for inflater
- Using the uncompressed length stored in the binary file to allocate appropriately sized inflater output array.
Improved readReplayData performance
- Eliminated unnecessary array copying
- Reusing ByteBuffer for UTF8 decode
Improved decompress performance by Reusing ByteBuffer for Base64 decode

Rogerstalker commented:
> This is a very interesting contribution and I was not aware of ByteBuffers or even its capability to go LittleEndian. I always wondered why we need Guava to do such a basic thing, well we were using the wrong thing as it seems.
> 
> I think the benefits are worth breaking the compatibility here. But we also need to adapt the moderator client later on.
> 
> So if we are breaking compatibility anyway we should go all in 😛 Since you are investigating and measuring I have a request: can you check if this also makes sense to apply to loading the data? There we still copy around data and that could be avoided.
> 
> This is a suggestion of ChatGPT, I didn't check it

I tried to improve its performance originally, since decompressing takes a significant amount of time, but it looked like the used libs didn't support some things I needed. I dug a bit deeper now, turns out I could manage to reduce the unnecessary copying. See latest commit.

I don't have such detailed measurements now for each change, since I made a ton of samplings and lost track at some point. 😄 
I'd say that the two changes that had the biggest impact were the elimination of array copying for the inflater and using the uncompressed length to allocate properly sized arrays. I also tried using ByteBuffer.allocateDirect and ByteBuffer.allocate to pass a ByteBuffer to the inflater instead of the byte array, but all three of these options were essentially within my measurement error range (hence the ton of samplings). I'd give a slight edge to using the array and wrapping it afterwards, it seemed slightly faster. I attribute this to the inflater's need to operate the ByteBuffer and its properties like limit, pos, capacity instead of just pushing data to the array.

So here are the most important measurements:
[Improved decompress.zip](https://github.com/user-attachments/files/22267383/Improved.decompress.zip)

Baseline
**ByteBuffer full final parallel.nps:** com.faforever.commons.replay.ReplayLoaderTableParserTest.ExecuteParallel () 13 299 ms

Improved decompress parallel
**ByteBuffer full final improved decompress.nps** java.util.concurrent.AbstractExecutorService.invokeAll ()	8 601 ms

I also made new measurements for the client with some minor changes to adjust to the common changes. I'll push these to the client PR if this is ok.

New baseline made with the non-improved decompress, since last 100 local replays changed.
**Client improved single buffer single changeload fix parallel 1stream pr baseline 100.nps:** com.faforever.client.replay.ReplayService.loadLocalReplayPage ()	4 502 ms

Improved decompress
**Client improved single buffer single changeload fix parallel 1stream improved decompress 100.nps:** com.faforever.client.replay.ReplayService.loadLocalReplayPage ()	2 901 ms

Rogerstalker commented:
I am not sure about the ZSTD compression though, since all local replays use QTCOMPRESS. Based on some of the issues/commits I found I would have expected all replay files to use ZSTD. https://github.com/FAForever/downlords-faf-client/issues/1374, https://github.com/FAForever/faf-aio-replayserver/pull/17

Brutus5000 commented:
@Rogerstalker All replays from the server have been retroactively converted to Zstd, because there disk size is limited. Locally I am not sure what is used. Either all Qtcompress or only older ones and the rest Zstd.

Brutus5000 commented:
Since this is a common library and other people might use it, I would rather keep `commandsPerMinuteByPlayer`. Is there any performance reason to remove it?

Rogerstalker commented:
> Since this is a common library and other people might use it, I would rather keep `commandsPerMinuteByPlayer`. Is there any performance reason to remove it?

You can check In the **original.nps** measurement: java.util.HashMap.computeIfAbsent ()	899 ms. Initially I wanted to improve its performance, by using an array instead of the outer Map. This would be possible, because we already know the number of players, that is why I made `playerIdsByName` a class member. But then I saw that `commandsPerMinuteByPlayer` is a private member and it is never used aside from adding data to it, so I thought it might be some now unused, leftover code and removed it.

Rogerstalker commented:
Ah I see it had a getter. I tried to check if it was used elsewhere: https://github.com/search?q=org%3AFAForever%20commandsPerMinuteByPlayer&type=code. I don't think it was. It had a performance impact, because was called a lot. We could add it back, but I don't see anyone using it and even if I could improve it, there would still be some impact.

Brutus5000 commented:
You are right. Should this ever be needed, we'd need to allow passing the events to a consumer passed as a constructor.
